ctapipe/io/magiceventsource.py

```diff
@@ -0,0 +1,1352 @@
+        optics = OpticsDescription.from_name('MAGIC')
+        geom = CameraGeometry.from_name('MAGICCam')
+        self.magic_tel_description = TelescopeDescription(optics=optics, camera=geom)
+        self.magic_tel_descriptions = {1: self.magic_tel_description, 2: self.magic_tel_description}
+        self.magic_subarray = SubarrayDescription('MAGIC', self.magic_tel_positions, self.magic_tel_descriptions)
+
+    @staticmethod
+    def is_compatible(file_mask):
+        """
+        This method checks if the specified file mask corresponds
+        to MAGIC data files. The result will be True only if all
+        the files are of ROOT format and contain an 'Events' tree.
+
+        Parameters
+        ----------
+        file_mask: str
+            A file mask to check
+
+        Returns
+        -------
+        bool:
+            True if the masked files are MAGIC data runs, False otherwise.
+
+        """
+
+        is_magic_root_file = True
+
+        file_list = glob.glob(file_mask)
+
+        for file_path in file_list:
+            try:
+                import uproot
+
+                try:
+                    with uproot.open(file_path) as input_data:
+                        if 'Events' not in input_data:
+                            is_magic_root_file = False
+                except ValueError:
+                    # uproot raises ValueError if the file is not a ROOT file
+                    is_magic_root_file = False
+                    pass
+
+            except ImportError:
+                if re.match('.+_m\d_.+root', file_path.lower()) is None:
+                    is_magic_root_file = False
+
+        return is_magic_root_file
+
+    @staticmethod
+    def _get_run_number(file_name):
+        """
+        This internal method extracts the run number from
+        the specified file name.
+
+        Parameters
+        ----------
+        file_name: str
+            A file name to process.
+
+        Returns
+        -------
+        int:
+            A run number of the file.
+        """
+
+        mask = ".*\d+_M\d+_(\d+)\.\d+_.*"
+        parsed_info = re.findall(mask, file_name)
+
+        try:
+            run_number = int(parsed_info[0])
+        except IndexError:
+            raise IndexError('Can not identify the run number of the file {:s}'.format(file_name))
+
+        return run_number
+
+    def _set_active_run(self, run_number):
+        """
+        This internal method sets the run that will be used for data loading.
+
+        Parameters
+        ----------
+        run_number: int
+            The run number to use.
+
+        Returns
+        -------
+
+        """
+
+        input_path = '/'.join(self.input_url.split('/')[:-1])
+        this_run_mask = input_path + '/*{:d}*root'.format(run_number)
+
+        run = dict()
+        run['number'] = run_number
+        run['read_events'] = 0
+        run['data'] = MarsDataRun(run_file_mask=this_run_mask)
+
+        return run
+
+    def _generator(self):
```

Comment:
I think this should be removed as it just calls another generator

Reply:
no. It could also overwrite `__iter__` to call  `iter_stereo_events()`

Reply:
Requiring a generator does not make sense for every eventsource, imho.

Reply:
An EventSource could also be a simple iterable, not a generator

Reply:
> An EventSource could also be a simple iterable, not a generator

Sure, but the iterable should act as a generator. ctapipe works on a single event container at a time...the functionality of the EventSource is that it returns one event container at a time.

Reply:
> And what is saved by requesting this eventsource subclass to override __iter__ instead of defining _generator?

That we use the standard python why? Principal of least suprise.

Reply:
> In what circumstance would an eventsource not utilise a generator? 

A generator is something very specific. An EventSource imho only needs to be an iterable of events. 

In the most simple case, i would say an EventSource only needs to implement the iterator protocol.

Reply:
This is more a question of the EventSource design, not how the MAGIC EventSource should be implemented. It would be more appropriate to discuss this separately.

When designing the EventSource, it was requested that the base EventSource class handles the max_events check, therefore avoiding every EventSource developer having to define it. I think this was a good suggestion, and the best way I saw to do that was to have the max_events checked in `__iter__`. This therefore required a separate function to `__iter__` that could be overridden in the EventSource subclass - hence `_generator`.

If you have improvements to suggest to EventSource then it would be more helpful to open a PR or issue on the subject. The changes you are requesting here go against the current EventSource design.